File: Data Structures & Algorithms/permutations/submission-3.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        
        def backtrack(st: int):
            if st == len(nums):
                ans.append(nums[:])
                return
            
            for i in range(st, len(nums)):
                nums[st], nums[i] = nums[i], nums[st]
                backtrack(st + 1)
                nums[st], nums[i] = nums[i], nums[st]

        backtrack(0)
        return ans

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.java
class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length, pre = 1, suf = 1, ans = Integer.MIN_VALUE;
        for(int i = 0; i < n; ++i) {
            pre *= nums[i];
            suf *= nums[n - i - 1];

            ans = Math.max(ans, Math.max(pre, suf));
            if(pre == 0) pre = 1;
            if(suf == 0) suf = 1;
        }
        return ans;
    }
}
